tower-http/src/cors.rs

```diff
@@ -0,0 +1,749 @@
+//!
+//! let request = Request::builder()
+//!     .header(header::ORIGIN, "https://example.com")
+//!     .body(Body::empty())
+//!     .unwrap();
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(
+//!     response.headers().get(header::ACCESS_CONTROL_ALLOW_ORIGIN).unwrap(),
+//!     "*",
+//! );
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+
+use futures_core::ready;
+use http::{
+    header::{self, HeaderName, HeaderValue},
+    request::Parts,
+    Method, Request, Response, StatusCode,
+};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`Cors`] which adds headers for [CORS][mdn].
+///
+/// See the [module docs](crate::cors) for an example.
+///
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+#[derive(Debug, Clone)]
+pub struct CorsLayer {
+    allow_credentials: Option<HeaderValue>,
+    allow_headers: HeaderValue,
+    allow_methods: HeaderValue,
+    allow_origin: AnyOr<Origin>,
+    expose_headers: Option<HeaderValue>,
+    max_age: Option<HeaderValue>,
+}
+
+const DEFAULT_METHODS: &str = "GET,POST,OPTIONS";
+const WILDCARD: &str = "*";
+
+impl CorsLayer {
+    /// Create a new `CorsLayer`.
+    pub fn new() -> Self {
+        Self {
+            allow_credentials: None,
+            allow_headers: WILDCARD.parse().unwrap(),
+            allow_methods: DEFAULT_METHODS.parse().unwrap(),
+            allow_origin: AnyOr(AnyOrInner::Any),
+            expose_headers: None,
+            max_age: None,
+        }
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Credentials`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    ///
+    /// let layer = CorsLayer::new().allow_credentials(true);
+    /// ```
+    ///
+    /// By default the header will not be set.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+    pub fn allow_credentials(mut self, allow_credentials: bool) -> Self {
+        self.allow_credentials = Some(allow_credentials.to_string().parse().unwrap());
+        self
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Headers`][mdn] header.
+    ///
+    /// ```
+    /// use tower_http::cors::CorsLayer;
+    /// use http::header::HeaderValue;
+    ///
+    /// let layer = CorsLayer::new().allow_headers(vec![
+    ///     "*".parse().unwrap(),
+    /// ]);
+    /// ```
+    ///
+    /// By default the header will be set to `*`.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers
+    pub fn allow_headers<I>(mut self, headers: I) -> Self
+    where
+        I: IntoIterator<Item = HeaderValue>,
```

Comment:
This might need to be `IntoIterator<Item = HeaderName>`. I was trying to do something like this, which resulted in an error (`expected struct HeaderName, found struct HeaderValue`):

```
use axum::http::header;

CorsLayer::new().allow_headers(vec![
  header::ORIGIN,
  header::CONTENT_LENGTH,
  header::CONTENT_TYPE,
  "X-Token".parse().unwrap()
])
```

 We previously used actix-cors, and [it seems like they do `Item = TryInto<HeaderName>`](https://docs.rs/actix-cors/0.5.4/src/actix_cors/builder.rs.html#288-289), which feels nicer since I don't have to do `.parse().unwrap()` for custom headers.

Reply:
Oh yeah! Thanks for noticing this! Its just a scary typo. I guess we need to add some more doc tests to make sure things actually work.